### PR TITLE
Add more granular system sets for state transition schedule ordering

### LINKED ISSUE #13711: Ordered `OnExit` and `OnEnter`
## What problem does this solve or what need does it fill?

Currently `OnExit` and `OnEnter` aren't ordered in relation to the state dependency graph.
This is a problem when parent state transition prepares e.g. resources for substates, but the `OnEnter(Substate)` runs first.

## What solution would you like?

- Order `OnExit` to run from graph leaf-states to root-states.
- Keep `OnTransition` in arbitrary order.
- Order `OnEnter` to run from graph root-states to leaf-states.

This can either be done by adding `after` conditions during state installation or as part of `StateRegistry`, a separate feature.

## What alternative(s) have you considered?

None that seem reasonable.

## Additional context

This could be considered a bug to some degree.

[Working example](https://github.com/MiniaczQ/states-ui-test), entering `Settings` will crash or not, depending on the system order that was decided during startup.

### PR DESCRIPTION
**Pull request by MiniaczQ**
# Objective

Fixes #13711 

## Solution

Introduce smaller, generic system sets for each schedule variant, which are ordered against other generic variants:
- `ExitSchedules<S>` - For `OnExit` schedules, runs from leaf states to root states.
- `TransitionSchedules<S>` - For `OnTransition` schedules, runs in arbitrary order.
- `EnterSchedules<S>` - For `OnEnter` schedules, runs from root states to leaf states.

Also unified `ApplyStateTransition<S>` schedule which works in basically the same way, just for internals.

## Testing

- One test that tests schedule execution order

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
We should collaborate on either this or #13724: they seem very similar :)

File: crates/bevy_state/src/state/mod.rs [hunk 1]
```diff
@@ -619,4 +619,84 @@ mod tests {
...
+        let mut world = World::new();
+        setup_state_transitions_in_world(&mut world, None);
+        EventRegistry::register_event::<StateTransitionEvent<SimpleState>>(&mut world);
+        EventRegistry::register_event::<StateTransitionEvent<SubState>>(&mut world);
+        EventRegistry::register_event::<StateTransitionEvent<TestComputedState>>(&mut world);
+        world.insert_resource(State(SimpleState::B(true)));
+        world.init_resource::<State<SubState>>();
+        world.insert_resource(State(TestComputedState::BisTrue));
+        let mut schedules = world.remove_resource::<Schedules>().unwrap();
+        let mut apply_changes = Schedule::new(StateTransition);
```
Comment (lee-orr) [inline crates/bevy_state/src/state/mod.rs:637]: This overrides the schedule that gets created by `setup_state_transitions_in_world`.
```suggestion
        let mut apply_changes = schedules.get_mut(StateTransition).unwrap();
```

File: crates/bevy_state/src/state/mod.rs [hunk 2]
```diff
@@ -619,4 +619,84 @@ mod tests {
             1
         );
     }
+
+    #[derive(Resource, Default, Debug)]
+    struct TransitionTracker(Vec<&'static str>);
+
```
Comment (lee-orr) [inline crates/bevy_state/src/state/mod.rs:625]: For this test, I think it'd be ideal to have the computed state depend on both the main & sub states, so that there is no ambiguity in the desired scheduling. So I'd recommend adding a computed state for this:

```suggestion


    #[derive(PartialEq, Eq, Debug, Hash, Clone)]
    enum TransitionTestingComputedState {
        IsA,
        IsBAndEven,
        IsBAndOdd
    }

    impl ComputedStates for TransitionTestingComputedState {
        type SourceStates = (Option<SimpleState>, Option<SubState>);

        fn compute(sources: (Option<SimpleState>, Option<SubState>)) -> Option<Self> {
            match sources {
                (Some(simple), sub) => {
                    if simple == SimpleState::A
                    {
                        Some(Self::IsA)
                    } else if sub == Some(SubState::One) {
                        Some(Self::IsBAndOdd)
                    } else if sub == Some(SubState::Two) {
                        Some(Self::IsBAndEven)
                    } else {
                        None
                    }
                }
                _ => None,
            }
        }
    }
```

File: crates/bevy_state/src/state/mod.rs [hunk 3]
```diff
@@ -619,4 +619,84 @@ mod tests {
...
+    #[test]
+    fn check_transition_orders() {
+        let mut world = World::new();
+        setup_state_transitions_in_world(&mut world, None);
+        EventRegistry::register_event::<StateTransitionEvent<SimpleState>>(&mut world);
+        EventRegistry::register_event::<StateTransitionEvent<SubState>>(&mut world);
+        EventRegistry::register_event::<StateTransitionEvent<TestComputedState>>(&mut world);
```
Comment (lee-orr) [inline crates/bevy_state/src/state/mod.rs:632]:
```suggestion
        EventRegistry::register_event::<StateTransitionEvent<TransitionTestingComputedState>>(&mut world);
```

File: crates/bevy_state/src/state/mod.rs [hunk 4]
```diff
@@ -619,4 +619,84 @@ mod tests {
...
+        world.insert_resource(State(SimpleState::B(true)));
+        world.init_resource::<State<SubState>>();
+        world.insert_resource(State(TestComputedState::BisTrue));
```
Comment (lee-orr) [inline crates/bevy_state/src/state/mod.rs:635]:
```suggestion
        world.insert_resource(State(TransitionTestingComputedState::IsA));
```

File: crates/bevy_state/src/state/mod.rs [hunk 5]
```diff
@@ -619,4 +619,84 @@ mod tests {
...
+        let mut schedules = world.remove_resource::<Schedules>().unwrap();
+        let mut apply_changes = Schedule::new(StateTransition);
+        SimpleState::register_state(&mut apply_changes);
+        SubState::register_sub_state_systems(&mut apply_changes);
+        TestComputedState::register_computed_state_systems(&mut apply_changes);
```
Comment (lee-orr) [inline crates/bevy_state/src/state/mod.rs:640]:
```suggestion
        TransitionTestingComputedState::register_computed_state_systems(&mut apply_changes);
```

File: crates/bevy_state/src/state/mod.rs [hunk 6]
```diff
@@ -619,4 +619,84 @@ mod tests {
...
+        );
+        schedules.add_systems(
+            StateTransition,
+            register_transition("sub enter").in_set(EnterSchedules::<SubState>::default()),
+        );
+
+        schedules.add_systems(
+            StateTransition,
+            register_transition("computed exit")
+                .in_set(ExitSchedules::<TestComputedState>::default()),
```
Comment (lee-orr) [inline crates/bevy_state/src/state/mod.rs:679]:
```suggestion
                .in_set(ExitSchedules::<TransitionTestingComputedState>::default()),
```

File: crates/bevy_state/src/state/mod.rs [hunk 7]
```diff
@@ -619,4 +619,84 @@ mod tests {
...
+        );
+        schedules.add_systems(
+            StateTransition,
+            register_transition("computed transition")
+                .in_set(TransitionSchedules::<TestComputedState>::default()),
```
Comment (lee-orr) [inline crates/bevy_state/src/state/mod.rs:684]:
```suggestion
                .in_set(TransitionSchedules::<TransitionTestingComputedState>::default()),
```

File: crates/bevy_state/src/state/mod.rs [hunk 8]
```diff
@@ -619,4 +619,84 @@ mod tests {
...
+        );
+        schedules.add_systems(
+            StateTransition,
+            register_transition("computed enter")
+                .in_set(EnterSchedules::<TestComputedState>::default()),
```
Comment (lee-orr) [inline crates/bevy_state/src/state/mod.rs:689]:
```suggestion
                .in_set(EnterSchedules::<TransitionTestingComputedState>::default()),
```

File: crates/bevy_state/src/state/mod.rs [hunk 9]
```diff
@@ -619,4 +619,84 @@ mod tests {
...
+        );
+
+        world.insert_resource(schedules);
+
+        world.run_schedule(StateTransition);
+
+        let transitions = &world.resource::<TransitionTracker>().0;
+        println!("{:?}", transitions);
+        assert_eq!(transitions.len(), 9);
+        assert_eq!(transitions[0], "simple");
```
Comment (lee-orr) [inline crates/bevy_state/src/state/mod.rs:699]:
```suggestion
        
        assert_eq!(transitions[0], "computed exit");
        assert_eq!(transitions[1], "sub exit");
        assert_eq!(transitions[2], "simple exit");
        assert_eq!(transitions[3], "simple transition");
        assert_eq!(transitions[4], "sub transition");
        assert_eq!(transitions[5], "computed transition");
        assert_eq!(transitions[6], "simple enter");
        assert_eq!(transitions[7], "sub enter");
        assert_eq!(transitions[8], "computed enter");
```

File: crates/bevy_state/src/state/mod.rs [hunk 10]
```diff
@@ -619,4 +619,84 @@ mod tests {
...
+        assert_eq!(transitions[8], "simple");
```
Comment (lee-orr) [inline crates/bevy_state/src/state/mod.rs:700]:
```suggestion
```

**Review by lee-orr — APPROVED** [review]
Overall the PR is good - the main issue was in the test, so I'd recommend applying some changes there (I provided suggestions) - but once they're in I'll be happy with this.

File: crates/bevy_state/src/state/transitions.rs [hunk 11]
```diff
@@ -57,26 +57,56 @@ pub struct StateTransitionEvent<S: States> {
...
 /// These system sets are run sequentially, in the order of the enum variants.
 #[derive(SystemSet, Clone, Debug, PartialEq, Eq, Hash)]
-pub enum StateTransitionSteps {
+pub(crate) enum StateTransitionSteps {
     /// Parentless states apply their [`NextState<S>`].
     RootTransitions,
     /// States with parents apply their computation and [`NextState<S>`].
     DependentTransitions,
-    /// Exit schedules are executed.
+    /// Exit schedules are executed in leaf-root order
```
Comment (benfrankel) [inline crates/bevy_state/src/state/transitions.rs:65]: Tiny nit: I'd write this as `in leaf -> root order`, or `in leaf to root order`, or `from leaf to root`; otherwise it might read as a hyphen. Same for the other two doc comments.

File: crates/bevy_state/src/state/transitions.rs [hunk 12]
```diff
@@ -57,26 +57,56 @@ pub struct StateTransitionEvent<S: States> {
...
     TransitionSchedules,
-    /// Enter schedules are executed.
+    /// Enter schedules are executed in root-leaf order.
     EnterSchedules,
 }
 
-/// Defines a system set to aid with dependent state ordering
 #[derive(SystemSet, Clone, Debug, PartialEq, Eq, Hash)]
-pub struct ApplyStateTransition<S: States>(PhantomData<S>);
+/// Exit schedules are executed.
```
Comment (benfrankel) [inline crates/bevy_state/src/state/transitions.rs:74]: Something like "A system set that runs exit schedules for `S`" may be more clear. Same for the other two doc comments.

File: crates/bevy_state/src/state/transitions.rs [hunk 13]
```diff
@@ -57,26 +57,56 @@ pub struct StateTransitionEvent<S: States> {
...
+/// Enter schedules are executed.
+pub struct EnterSchedules<S: States>(PhantomData<S>);
+
+impl<S: States> Default for EnterSchedules<S> {
+    fn default() -> Self {
+        Self(Default::default())
+    }
+}
+
+/// Defines a system set to aid with dependent state ordering.
```
Comment (benfrankel) [inline crates/bevy_state/src/state/transitions.rs:103]: There are multiple system sets that help with dependent state ordering now, so this doc comment is vague.

File: crates/bevy_state/src/state/freely_mutable_state.rs [hunk 14]
```diff
@@ -17,27 +17,33 @@ use super::{take_next_state, transitions::*};
 pub trait FreelyMutableState: States {
     /// This function registers all the necessary systems to apply state changes and run transition schedules
     fn register_state(schedule: &mut Schedule) {
+        schedule.configure_sets((
+            ApplyStateTransition::<Self>::default()
+                .in_set(StateTransitionSteps::DependentTransitions),
```
Comment (benfrankel) [inline crates/bevy_state/src/state/freely_mutable_state.rs:22]: Why the change from `RootTransitions` to `DependentTransitions` for `FreelyMutableState` types?

**Review by benfrankel — COMMENTED** [review]
Some documentation nits & a question.

**Comment by benfrankel** [comment]
> Replaced `StateTransitionSteps` in favor of `ExitSchedules`, `TransitionSchedules`, `EnterSchedules`.

This in the changelog seems not entirely accurate. `StateTransitionSteps` is still around, and the new system sets provide internal ordering.

File: crates/bevy_state/src/state/freely_mutable_state.rs (hunk 14 above)
Comment (MiniaczQ) [reply crates/bevy_state/src/state/freely_mutable_state.rs:22]: Copy paste "error", this part of code is HEAVILY shared between all state traits, the only difference being the amount of dependencies (root states = none, sub/comp one [manual impl] or more [macros])

Comment (MiniaczQ) [reply crates/bevy_state/src/state/freely_mutable_state.rs:22]: I'll keep this and remove the `RootTransitions` instead

**Review by alice-i-cecile — APPROVED** [review]
Good tests, reasonable implementation. If you can get CI green I'll merge this tomorrow during the merge train.

**Review by benfrankel — APPROVED** [review]
I haven't reviewed the added test, but the implementation looks good.
